pkg/types/ccipocr3/config.go

```diff
@@ -78,3 +64,148 @@ type ObserverInfo struct {
+	}
+	return json.Marshal(temp)
+}
+
+type ConsensusObservation struct {
+	// FChain defines the FChain value for each chain. FChain is used while forming consensus based on the observations.
+	FChain map[ChainSelector]int `json:"fChain"`
+	// PricedTokens is a list of tokens that we want to submit price updates for.
+	PricedTokens []types.Account `json:"pricedTokens"`
+	// NodeSupportedChains is a map of oracle IDs to SupportedChains.
+	NodeSupportedChains P2PToSupportedChains `json:"nodeSupportedChains"`
+}
+
+func (c ConsensusObservation) Validate() error {
+	for _, inf := range c.NodeSupportedChains {
+		for ch := range inf.Supported.Iter() {
+			if _, ok := c.FChain[ch]; !ok {
+				return fmt.Errorf("fChain not set for chain %d", ch)
+			}
+		}
+	}
+	if len(c.PricedTokens) == 0 {
+		return fmt.Errorf("priced tokens not set, at least one priced token is required")
+	}
+
+	return nil
+}
+
+type HomeChainConfig struct {
+	// FChain defines the FChain value for each chain. FChain is used while forming consensus based on the observations.
+	FChain map[ChainSelector]int `json:"fChain"`
+	// NodeSupportedChains is a map of PeerIDs to SupportedChains.
+	NodeSupportedChains P2PToSupportedChains `json:"nodeSupportedChains"`
+}
+
+func (c *HomeChainConfig) Validate() error {
+	for _, inf := range c.NodeSupportedChains {
+		for ch := range inf.Supported.Iter() {
+			if _, ok := c.FChain[ch]; !ok {
+				return fmt.Errorf("fChain not set for chain %d", ch)
+			}
+		}
+	}
+	return nil
+}
+func (c *HomeChainConfig) GetFChain(chain ChainSelector) int {
+	return c.FChain[chain]
+}
+
+func (c *HomeChainConfig) IsSupported(node P2PID, chain ChainSelector) bool {
+	supportedChains, ok := c.NodeSupportedChains[node]
+	if !ok {
+		return false
+	}
+	return supportedChains.IsSupported(chain)
+}
+
+func (c *HomeChainConfig) GetSupportedChains(node P2PID) mapset.Set[ChainSelector] {
+	supportedChains, ok := c.NodeSupportedChains[node]
+	if !ok {
+		return mapset.NewSet[ChainSelector]()
+	}
+	return supportedChains.Supported
+}
+
+type SupportedChains struct {
+	Supported mapset.Set[ChainSelector] `json:"supported"`
+}
+
+func (sc *SupportedChains) IsSupported(chain ChainSelector) bool {
+	return sc.Supported.Contains(chain)
+}
+
+func (sc *SupportedChains) UnmarshalJSON(data []byte) error {
+	temp := struct {
+		Supported []ChainSelector `json:"supported"`
+	}{}
+
+	if err := json.Unmarshal(data, &temp); err != nil {
+		return err
+	}
+
+	sc.Supported = mapset.NewSet[ChainSelector]()
+	for _, ch := range temp.Supported {
+		sc.Supported.Add(ch)
+	}
+
+	return nil
+}
+
+func (sc SupportedChains) MarshalJSON() ([]byte, error) {
+	supportedSlice := make([]ChainSelector, 0, sc.Supported.Cardinality())
+	for ch := range sc.Supported.Iter() {
+		supportedSlice = append(supportedSlice, ch)
+	}
+
+	// Define a temporary struct with a slice for Supported
+	temp := struct {
+		Supported []ChainSelector `json:"supported"`
+	}{
+		Supported: supportedSlice,
```

Comment:
Is this deterministically generated in the same way all the time? If this is using a `map` under the hood iteration order is undefined so you might get different lists on different nodes and hence consensus failure.

If there is a risk of this we should `sort` the slice first before plugging it into this struct.

Reply:
very good observation

Reply:
We talked about this, the plugin is already using a Set as well, comparing sets is easy using the library and doesn't need to be in a specific order or anything.